### ISSUE TITLE
Wrong service port name for imap in jail.conf  if you are using Nftables.

### ISSUE THREAD

**Issue by 1RUSUA1** [fail2ban/fail2ban#2744]
### Environment:
- Fail2Ban version: 1.0.1.dev1
- OS, including release name/version: 
         Debian GNU/Linux 10 (buster)
         Linux 4.19.0-8-cloud-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) x86_64 GNU/Linux
- [ ] Fail2Ban installed via OS/distribution mechanisms
- [X] You have not applied any additional foreign patches to the codebase
- [x] Some customizations were done to the configuration (provide details below is so)

### The issue:
If you using nftables, your jail.conf has wrong port name **imap** in jail configs. Nftables doesn't recognize port named **imap**. It needs **imap2** name.

Proof from log:
```
nft add rule ip fail2ban input $proto dport \{ 110,995,imap,imaps,465 \} ip saddr @addr-set-dovecot drop
done
[26564]: ERROR   7f253c0b05f0 -- stderr: 'Error: Could not resolve service: Servname not found in nft services list'
 [26564]: ERROR   7f253c0b05f0 -- stderr: 'add rule ip fail2ban input tcp dport { 110,995,imap,imaps,465 } ip saddr @addr-set-dovecot drop'
[26564]: ERROR   7f253c0b05f0 -- stderr: '                                               ^^^^'
[26564]: ERROR   7f253c0b05f0 -- returned 1
```

After I changed in jail.conf **imap** to **imap2** - nftables works good.

**Comment by asy972** [comment]
What is contained in /etc/services ? Show the output of the `grep "143/" /etc/services` command.

**Comment by sebres** [comment]
The issue is - `/etc/services` depends on distribution and its version extremely.
We cannot provide valid jail.conf for every version.
@fail2ban/maintainers we could move it to distribution related includes... but I guess it should also consider a version.

**Comment by 1RUSUA1** [comment]
Why don't you use numeric port instead of write it's name? This should solve such problem. Or not?

**Comment by opoplawski** [comment]
Why on earth is getservbyname() not working for the alias like it should?  See also https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=959989

**Comment by sebres** [comment]
> Why don't you use numeric port instead of write it's name? 

Because it is de facto standard, that one uses alias instead of number. This way it is meaningful (so more informative) and one doesn't need to change it in 100 different config files if it gets changed (customized).
But of course you could rewrite it by yourself in your `jail.local`.

**Comment by SpareSimian** [comment]
See https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=902752

**Comment by SpareSimian** [comment]
I don't think nft accepts a service name. I don't see a call to getservbyname in its sources. So fail2ban will need to to the lookup and pass the number to it.

https://git.netfilter.org/nftables/

**Comment by sebres** [comment]
> I don't think nft accepts a service name.

It does. Just for some reason not for an alias, so `imap2` works whether `imap` (which is an alias) don't.
```
$ nft add rule inet f2b-test f2b-chain tcp dport \{ imap \} ip saddr @addr-set-test reject
Error: Could not resolve service: Servname not found in nft services list
add rule inet f2b-test f2b-chain tcp dport { imap } ip saddr @addr-set-test reject
                                             ^^^^
$ nft add rule inet f2b-test f2b-chain tcp dport \{ imap2 \} ip saddr @addr-set-test reject
<no errors>
```

> I don't see a call to getservbyname in its sources

And this is probably the reason. If they provided own implementation of `getservbyname` in nft, then it is done wrong (without to consider the aliases).

> So fail2ban will need to to the lookup and pass the number to it.

Well, as already said service names will be accepted by nftables. As for imap vs. imap2, normally it is matter of maintainers of certain distribution (and version). But since our primary target are Debian systems, we could indeed replace imap with imap2 (worse is that this will affect all other distributions and versions), but nftables is not yet default banning action.
Just as Orion correctly said (in https://github.com/fail2ban/fail2ban/issues/2744#issuecomment-646924315), it is strange that an alias doesn't accepted at all - rather it is a bug of nftables and must be fixed there.

Here is a related test for iptables:
```
$ cat /etc/services | grep imap2
imap2           143/tcp         imap            # Interim Mail Access P 2 and 4

$ iptables -w -N f2b-test
$ iptables -w -I f2b-test -p tcp -m multiport --dports imap -d 192.0.2.1 -j REJECT
$ iptables -w -I f2b-test -p tcp -m multiport --dports imap2 -d 192.0.2.2 -j REJECT

$ iptables -nL f2b-test
Chain f2b-test (0 references)
target     prot opt source               destination
REJECT     tcp  --  0.0.0.0/0            192.0.2.1            multiport dports 143 reject-with icmp-port-unreachable
REJECT     tcp  --  0.0.0.0/0            192.0.2.2            multiport dports 143 reject-with icmp-port-unreachable

$ iptables -w -F f2b-test
$ iptables -w -X f2b-test
```

**Comment by SpareSimian** [comment]
Thanks, @sebres, I guess I need to stare at the nft source some more. Maybe the lookup is done in a library. I'm using CentOS 7 where imap is the primary service name and imap2 is the alias. (nft isn't used until CentOS 8.) Proper lookups on the alias should fix this for all distros. Alas, I can't find an issue tracker for nft on the netfilter site. So maybe open a ticket on Debian's nft package.

**Comment by SpareSimian** [comment]
Debian bug tracker says it's fixed in nftables/0.9.7-1. https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=959989

**Comment by SpareSimian** [comment]
For curious lurkers, I did track down the call to getservbyname. It's invoked internally by glibc's getaddrinfo (the preferred entrypoint in new code), which is called by nftables's inet_service_type_parse in datatype.c. I couldn't find a version of these that was obviously ignoring the aliases. Without a failing system to inspect, I don't see a way to investigate further. We'd need to see the exact versions of /etc/services, nftables, and glibc so we can find the source code to this stack. But it's clearly a 3rd party issue.
